fastpair/rust/demo/rust/src/decoder.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use std::fs;
+
+use serde::Deserialize;
+
+use bluetooth::ServiceData;
+
+use crate::advertisement::ModelId;
+
+/// Unit struct providing parsing operations for Fast Pair advertisements.
+pub(crate) struct FpDecoder;
+
+/// Holds Fast Pair device information parsed from JSON.
+#[derive(Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub(crate) struct DeviceInfo {
+    image_url: String,
+    name: String,
+}
+
+/// Holds top-level Fast Pair information parsed from JSON. See `local`
+/// directory for format.
+#[derive(Deserialize)]
+struct JsonData {
+    device: DeviceInfo,
+}
+
+impl FpDecoder {
+    /// Retrieve device information for the provided Model ID. Currently,
+    /// this information is saved locally. In the future, this should instead
+    /// be retrieved from a remote server and cached.
+    /// b/294456411
+    pub(crate) fn get_device_info_from_model_id(
+        model_id: &ModelId,
+    ) -> Result<DeviceInfo, anyhow::Error> {
+        let file_path = format!("./local/{model_id}.json");
+        let contents = fs::read_to_string(file_path).expect("Couldn't find or open file.");
+
+        let model_info: JsonData = serde_json::from_str(&contents)?;
+
+        Ok(model_info.device)
+    }
+
+    /// Retrieve the model ID length from the first byte of the 16-bit UUID
+    /// service data, for situations where service data length > 3.
+    fn get_id_length(id: u8) -> usize {
+        static HEADER_LENGTH_BITMASK: u8 = 0b00011110;
+        static HEADER_LENGTH_OFFSET: u8 = 1;
+
+        usize::from((id & HEADER_LENGTH_BITMASK) >> HEADER_LENGTH_OFFSET)
+    }
+
+    /// Retrieve the Fast Pair device model ID from a service data payload.
+    /// * Length < 3: invalid payload
+    /// * Length == 3: entire payload is the model ID
+    /// * Length > 3: first byte specifies the length in bytes of the model ID.
+    pub(crate) fn get_model_id_from_service_data<U: Copy>(
```

Comment:
parsing could be simplified by using a parser library like `nom`. For instance, `nom` supports length prefixes built in: https://docs.rs/nom/latest/nom/multi/fn.length_value.html

Reply:
I wrote this parsing functionality to mimic the C++ code, but later learned from my team that no devices currently use the more complicated length prefix, and the public API documentation for Fast Pair doesn't mention it either. I still kept the implementation to save someone time in the future, but since there's a strictly better way of parsing I'll just delete my implementation and put down Buganizer b/294453912.